docs/proposals/PROPOSAL_B.md

```diff
@@ -0,0 +1,405 @@
+
+The compatibility spec must be interpreted by a dedicated software.
+
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.oci.image-compatibility.v1",
+  "config": {
+    "mediaType": "application/vnd.oci.empty.v1+json",
+    "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
+    "size": 2
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.oci.image-compatibility.spec.v1+json",
+      "digest": "sha256:4a47f8ae4c713906618413cb9795824d09eeadf948729e213a1ba11a1e31d052",
+      "size": 1710
+    }
+  ],
+  "annotations": {
+    "oci.opencontainers.image.created": "2024-01-02T03:04:05Z"
+  }
+}
+```
+
+### Relation of the Artifact to Image Manifest
+
+```text
+ Image Index                                                 
+ +----------------------------------------------------------+
+ |                                                          |
+ |  Image Manifest A             Image Manifest B           |
+ |  +-----------------------+    +-----------------------+  |
+ |  |platform:              |    |platform:              |  |
+ |  |  archictecture: amd64 |    |  archictecture: arm   |  |
+ |  |  os: linux            |    |  os: linux            |  |
+ |  |                       |    |                       |  |
+ |  |digest: abc123         |    |digest: def456         |  |
+ |  +--------------------|--+    +--------------------|--+  |
+ |                       |                            |     |
+ +-----------------------|----------------------------|-----+
+                         |                            |      
+                         |                            |      
+    Artifact A           |       Artifact B           |      
+    +--------------------|--+    +--------------------|--+   
+    |                    |  |    |                    |  |   
+    | Subject Descriptor |  |    | Subject Descriptor |  |   
+    | +------------------|-+|    | +------------------|-+|   
+    | | digest: abc123     ||    | | digest: def456     ||   
+    | +--------------------+|    | +--------------------+|   
+    |                       |    |                       |   
+    +-----------------------+    +-----------------------+   
+```
+
+The design imposes a strong 1 to 1 relationship between the compatibility artifact and the image.
+That allows users to independently release artifacts and attach compatibility to the already existing images.
+
+The disadvantage of this solution is that you cannot create compatibility that points to multiple images having the same requirements.
+
+### Artifact Discovery
+
+If a registry supports referrers API, it should be used for the artifact discovery.
+
+Otherwise, the fallback method on the client said shall be implemented.
+The client:
+
+1. Fetches the image index.
+1. Iterates over descriptors to check if it's compatibility artifact type.
```

Comment:
Does the search stop on the first matching artifactType? What happens if there are multiple compatibility specs for an image?

Reply:
> Does the search stop on the first matching artifactType?

It should stop on the first matching artifactType. 

> What happens if there are multiple compatibility specs for an image?

Either, it should return an error or get the first one matching.
This is debatable but I am in favor of the first option to keep the following point valid:
["The design imposes a strong 1 to 1 relationship between the compatibility artifact and the image."](https://github.com/opencontainers/wg-image-compatibility/pull/6/files#diff-644786f78aa35025b6d4493a27fcc3d34c7a50df73c81c135daff93025a620c8R71)

Reply:
After speaking to @ChaoyiHuang internally. We decided to check every arttifact and pull the latest based on the creation time.

Reply:
> What happens if there are multiple compatibility specs for an image?

I hope there isn't a use case for this - if there are different needs for compatibility I'd want it represented in the same artifact, one source of truth to be decided how to act upon by some tool. That said, we can't control what people do, and in that light choosing that latest/newest is a good strategy.

Reply:
> I hope there isn't a use case for this

There is no user use case for that, however, it can happen that in certain cases we end up having multiple artifacts reffering to the same image. For instance after an update of the artifact you may have the old and new one pointing to the same image. If users are not careful enough and they push a new version of the same artifact with a different tag etc.